{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "cb8a13ce",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "The DREAM R package provides a comprehensive set of functionalities to explore and predict drug combinations based on gene expression and perturbation data. \n",
    "In this vignette, we outline the main steps of the analysis and explain how to utilize the package.\n",
    "\n",
    "### Analysis Workflow\n",
    "\n",
    "The analysis workflow consists of the following steps:\n",
    "  1. Building a Disease Network: The first step involves constructing a disease network based on differential gene expression profiles. \n",
    "  This network captures the interactions between genes and serves as the foundation for subsequent analyses.\n",
    "\n",
    "  2. Identifying Druggable Target Genes: Next, we identify the druggable target genes within the disease network.\n",
    "     By pinpointing specific genes that can be targeted by drugs, we gain valuable insights into potential therapeutic opportunities.\n",
    "\n",
    "  3. Building Drug Similarity Matrices: To assess the similarity between drugs, we construct multiple drug distance metrics based on molecular and chemical aspects.\n",
    "     These distances enable us to evaluate the relationships and potential combinations among different drugs.\n",
    "\n",
    "  4. Predicting Drug Combinations: Leveraging the disease network, druggable target genes, and drug distances, we utilize a multi-objective genetic algorithm to predict and prioritize potential drug combinations.\n",
    "     This step is crucial in identifying drug interactions and optimizing therapeutic interventions.\n",
    "\n",
    "### Required Inputs\n",
    "\n",
    "To perform the analysis, the following inputs are required:\n",
    "\n",
    "  1. Gene Expression Experiments: At least two sets of gene expression experiments representing different experimental conditions are required.\n",
    "     These experiments should capture the transcriptional responses under various conditions or treatments.\n",
    "\n",
    "  2. Transcriptional Perturbation Profiles: Perturbation profiles derived from drug treatments are essential for identifying druggable targets and predicting drug combinations.\n",
    "  These profiles provide valuable information on how genes are affected by specific drugs.\n",
    "\n",
    "### Warning: Execution time\n",
    "Please note that some of the code chunks in this analysis may take a significant amount of time to execute, especially when dealing with large datasets or performing computationally intensive tasks.\n",
    "\n",
    "To minimize waiting time, we provide intermediate output files for certain steps [here](https://github.com/fhaive/dream/releases/download/public/intermediate.tar.gz).\n",
    "These files contain pre-computed results that can be loaded directly, without the need to re-run time-consuming code chunks.\n",
    "\n",
    "### Python setup\n",
    "\n",
    "Some steps of the analysis require python to run, so the path to the interpreter must be provided.\n",
    "The dream package comes with both python and R environments with all the necessary dependencies listed in the `environment.yml` file.\n",
    "To get the path of the python interpreter it is sufficient to run these commands in a bash shell:\n",
    "```\n",
    "conda run -n dream which python\n",
    "```\n",
    "Copy and paste this path in the chunk below as the value of the variable `python_bin`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4adf0162",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "python_bin = \"/opt/mambaforge/bin/python\"\n",
    "library(DREAM)\n",
    "\n",
    "# setup data paths, the root directory can be overridden by changing the following instruction\n",
    "rootdir = \"/workspace\"\n",
    "data_path = file.path(rootdir, \"data\")\n",
    "intermediate_path = file.path(rootdir, \"intermediate\")\n",
    "\n",
    "dir.create(data_path, showWarnings=FALSE)\n",
    "# this folder containes intermediate files, they can either be generated by the notebook, or they can be downloaded from [...] to save computation time\n",
    "dir.create(intermediate_path, showWarnings=FALSE)\n",
    "drug_expr_path = file.path(intermediate_path, \"drug_expr\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "127efed3",
   "metadata": {},
   "source": [
    "# Generating Co-expression Disease Network for Atopic Dermatitis\n",
    "In this step, we will generate a co-expression disease network for Atopic Dermatitis (AD) based on informative genes identified as differentially expressed genes (DEGs).\n",
    "\n",
    "## Data Retrieval\n",
    "We retrieve the gene expression data for both lesional and non-lesional AD conditions. \n",
    "We will use the `download.file` function to download the data files from Zenodo. \n",
    "The URLs for the data files are:\n",
    "\n",
    "  Lesional AD data: [GE_Mic_AD_Pamr_MAARS.txt](https://zenodo.org/record/7962332/files/GE_Mic_AD_Pamr_MAARS.txt?download=1)\n",
    "  Non-lesional AD data: [GE_Mic_AD_Pamr_nl_MAARS.txt](https://zenodo.org/record/7962332/files/GE_Mic_AD_Pamr_nl_MAARS.txt?download=1)\n",
    "\n",
    "The code for downloading the data files is as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b6120d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(timeout=3600)\n",
    "ad_gex_url <- \"https://zenodo.org/record/7962332/files/GE_Mic_AD_Pamr_MAARS.txt?download=1\"\n",
    "ad_gex_nl_url <- \"https://zenodo.org/record/7962332/files/GE_Mic_AD_Pamr_nl_MAARS.txt?download=1\"\n",
    "download.file(ad_gex_url, file.path(data_path, \"GE_Mic_AD_Pamr_MAARS.txt\"), method=\"wget\", )\n",
    "download.file(ad_gex_nl_url, file.path(data_path, \"GE_Mic_AD_Pamr_nl_MAARS.txt\"), method=\"wget\", )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79e4a28",
   "metadata": {},
   "source": [
    "## Data Preprocessing\n",
    "We define a function to process the downloaded gene expression data. \n",
    "This function reads the data file, retrieves the gene symbol information using Ensembl gene IDs, removes rows with missing symbols, and returns the processed expression data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a32baafb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "process_gene_expression_data <- function(data_file, ensembl_dataset) {\n",
    "  expression_data <- read.table(data_file, header = TRUE, sep = \"\\t\", quote = \"\")\n",
    "  \n",
    "  conversion_table <- biomaRt::getBM(\n",
    "    attributes = c(\"ensembl_gene_id\", \"hgnc_symbol\"),\n",
    "    filters = \"ensembl_gene_id\",\n",
    "    values = rownames(expression_data),\n",
    "    mart = ensembl_dataset\n",
    "  )\n",
    "  \n",
    "  conversion_table$hgnc_symbol[which(conversion_table$hgnc_symbol == \"\")] <- NA\n",
    "  \n",
    "  expression_data$symbols <- conversion_table$hgnc_symbol[match(rownames(expression_data), conversion_table$ensembl_gene_id)]\n",
    "  expression_data <- na.omit(expression_data)\n",
    "  rownames(expression_data) <- expression_data$symbols\n",
    "  expression_data$symbols <- NULL\n",
    "  \n",
    "  return(expression_data)\n",
    "}\n",
    "\n",
    "# Note: if these commands return a biomaRt server error try to rerun them\n",
    "ensembl <- biomaRt::useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\")\n",
    "\n",
    "ad_pamr <- process_gene_expression_data(file.path(data_path, \"GE_Mic_AD_Pamr_MAARS.txt\"), ensembl)\n",
    "ad_pamr_nl <- process_gene_expression_data(file.path(data_path, \"GE_Mic_AD_Pamr_nl_MAARS.txt\"), ensembl)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c74c082",
   "metadata": {
    "editable": true,
    "lines_to_next_cell": 0,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "Next, we identify the common genes present in both lesional and non-lesional AD data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c368abb5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "common_genes <- intersect(rownames(ad_pamr), rownames(ad_pamr_nl))\n",
    "ad_pamr <- ad_pamr[which(rownames(ad_pamr) %in% common_genes),]\n",
    "ad_pamr_nl <- ad_pamr_nl[which(rownames(ad_pamr_nl) %in% common_genes),]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06c8bf13",
   "metadata": {},
   "source": [
    "Finally, we select the informative genes (differentially expressed genes).\n",
    "We will use the `get_informative_genes` function, passing the combined data frame and the grouping information as parameters. \n",
    "In this case, the grouping information represents whether a sample is from the lesional AD group or the non-lesional AD group. \n",
    "The `test` parameter specifies the statistical test to use for gene selection, and the `percentile` parameter determines the threshold for selecting the top informative genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97407196",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# this computation can be skipped by loading this intermediate file\n",
    "# inputinform <- readRDS(file.path(intermediate_path, \"inputinform.rds\"))\n",
    "all_samples_filtered <- cbind(ad_pamr, ad_pamr_nl)\n",
    "inputinform <- get_informative_genes(expr_mat = all_samples_filtered, group = as.factor(c(rep(\"AD\", ncol(ad_pamr)), rep(\"N\", ncol(ad_pamr_nl)))), test = \"wilcoxon + var.test\", percentile = 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43a9722f",
   "metadata": {},
   "source": [
    "## Consensus Matrix Generation\n",
    "To create the co-expression network, we will use the `get_ranked_consensus_matrix` function. \n",
    "This function takes the gene expression table (`inputinform[[1]]`) containing the informative genes as input. \n",
    "The `gx_table` parameter represents the gene expression data for the DEGs.\n",
    "\n",
    "Several parameters can be specified to configure the co-expression network generation, such as the methods for network inference (`iMethods`), the correlation/association estimation method (`iEst`), and the discretization method (`iDisc`).\n",
    "We use all the possible combinations of values in order to obtain a more robust consensus co-expression network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b134a99",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Note: Warning - This code chunk takes a significant amount of time to run.\n",
    "generatematrices=get_ranked_consensus_matrix(\n",
    "    gx_table=inputinform[[1]], \n",
    "    iMethods=c(\"clr\",\"aracne\",\"mrnet\"),\n",
    "    iEst=c(\"pearson\",\"spearman\",\"kendall\",\"mi.empirical\",\"mi.mm\",\"mi.shrink\",\"mi.sg\"),\n",
    "    iDisc=c(\"none\",\"equalfreq\",\"equalwidth\",\"globalequalwidth\"),\n",
    "    ncores=32,\n",
    "    debug_output=TRUE,\n",
    "    updateProgress=TRUE\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d74d633",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Edge Selection\n",
    "Next, we parse the edge rank matrix obtained from the consensus matrix and perform edge selection based on different strategies. \n",
    "The `parse_edge_rank_matrix` function is used for this purpose.\n",
    "\n",
    "The function takes the edge rank matrix (`edge_rank_matrix`) as input and specifies the edge selection strategy (`edge_selection_strategy`).\n",
    "Additionally, the `mat_weights` parameter determines the weight attribute used for edge selection, and `topN` specifies the number of top-ranked edges to select."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c83d0e1e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "rankMat.parsed <- parse_edge_rank_matrix(\n",
    "    edge_rank_matrix=generatematrices,\n",
    "    edge_selection_strategy=\"default\",\n",
    "    mat_weights=\"rank\",\n",
    "    topN=10,\n",
    "    debug_output=TRUE,\n",
    "    updateProgress=TRUE\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3622caa",
   "metadata": {},
   "source": [
    "We create an iGraph object from the symmetrical adjacency matrix. This object annotated with centrality attributes will be used in following steps of the analysis.\n",
    "Finally, we convert the iGraph object into a data frame format (`disease_network`) that represents the edges in the network.\n",
    "Each row of the data frame contains two genes representing an edge in the network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9930e0b6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# this computation can be skipped by loading this intermediate file\n",
    "# conGraph <- readRDS(file.path(intermediate_path, \"conGraph.rds\"))\n",
    "conGraph <- get_iGraph(rankMat.parsed$bin_mat)\n",
    "disease_network <- data.frame(as_edgelist(as.undirected(conGraph, mode=\"collapse\")))\n",
    "colnames(disease_network) <- c(\"gene1\", \"gene2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba7e9237",
   "metadata": {},
   "source": [
    "## Ranking Genes in the Disease Network\n",
    "We rank the genes in the disease network based on their differential expression. \n",
    "We use the `get_graph_named_ranklist` function, which takes an iGraph object (`conGraph`) and a list of ranked vectors as input.\n",
    "The `get_graph_named_ranklist` function assigns the ranked vectors as vertex attributes in the graph and calculates a ranked gene list using specified network attributes. \n",
    "The resulting ranked gene list is returned as a named list.\n",
    "In this case, the ranked vectors used for ranking the genes are `inputinform[[\"var_score\"]]` and `inputinform[[\"wilcox_score\"]]`. \n",
    "These vectors represent the scores of rankings of genes based on their variability and significance in the differential expression analysis.\n",
    "We convert the ranked gene list into a data frame format. \n",
    "Each row of the data frame contains a gene name and its corresponding rank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7477b886",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "graph_rank <- get_graph_named_ranklist(\n",
    "    Graph=conGraph,\n",
    "    ranked_vectors=list(inputinform[[\"var_score\"]],\n",
    "    inputinform[[\"wilcox_score\"]])\n",
    ")\n",
    "rank <- data.frame(gene=names(graph_rank), rank=graph_rank)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b99bdcd",
   "metadata": {},
   "source": [
    "# Druggability evaluation of the disease under investigation and computation of drug properties.\n",
    "\n",
    "## Downloading OpenTargets JSON File\n",
    "First, we download the OpenTargets JSON file from the specified URL (`opentargets_url`) using the `download.file` function. \n",
    "The downloaded file will be saved as \"./data/19.02_evidence_data.json.gz\". \n",
    "This file is in a compressed format (`Gzip`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9d6ce90",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Note: Warning - This code chunk takes a significant amount of time to run.\n",
    "opentargets_url <- \"http://ftp.ebi.ac.uk/pub/databases/opentargets/platform/19.02/output/19.02_evidence_data.json.gz\"\n",
    "download.file(opentargets_url, file.path(data_path, \"19.02_evidence_data.json.gz\"), method=\"wget\", )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b3be9b9",
   "metadata": {},
   "source": [
    "## Parsing OpenTargets JSON\n",
    "Next, we parse the downloaded OpenTargets JSON file using a custom convenience function called `parse_opentargets_json`.\n",
    "This function takes the JSON file (in Gzip format) as input and returns a parsed data frame with specific fields extracted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77baf6d6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Note: Warning - This code chunk takes a significant amount of time to run.\n",
    "# the intermediate result final_opentargets_parsed can be loaded with\n",
    "# final_opentargets_parsed <- readRDS(file.path(intermediate_path, \"final_opentargets_parsed.rds\"))\n",
    "final_opentargets_parsed = parse_opentargets_json(gzfile(file.path(data_path, \"19.02_evidence_data.json.gz\")))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dad5463",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Mapping OpenTargets Data on the Disease Network\n",
    "To identify druggable targets on the disease network, we map the OpenTargets data onto the disease network graph (`conGraph`).\n",
    "The OpenTargets data, which is stored in the `final_opentargets_parsed` data frame in the previous step, will be used for this mapping.\n",
    "\n",
    "The `map_opentargets_on_graph` function is designed to map OpenTargets data onto a graph.\n",
    "It takes the disease network graph and the parsed OpenTargets data as inputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ed4fbf8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "map_drugs_on_graph <- map_opentargets_on_graph(Graph = conGraph, parsed_opentargets = final_opentargets_parsed)\n",
    "map_drugs_on_graph <- unique(map_drugs_on_graph)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3b7f595",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Downloading OpenTargets JSON File\n",
    "We download the LINCS L1000 files from the list of URLs (`lincs_urls`) using the `download.file` function. \n",
    "The downloaded file will be saved as \"./data/GSE92742/*\". "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2a41120",
   "metadata": {
    "lines_to_next_cell": 2,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "lincs_urls <- list(\n",
    "  \"GSE92742_Broad_LINCS_pert_info.txt\"=\"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE92742&format=file&file=GSE92742%5FBroad%5FLINCS%5Fpert%5Finfo%2Etxt%2Egz\",\n",
    "  \"GSE92742_Broad_LINCS_sig_info.txt\"=\"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE92742&format=file&file=GSE92742%5FBroad%5FLINCS%5Fsig%5Finfo%2Etxt%2Egz\",\n",
    "  \"GSE92742_Broad_LINCS_gene_info_delta_landmark.txt\"=\"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE92742&format=file&file=GSE92742%5FBroad%5FLINCS%5Fgene%5Finfo%5Fdelta%5Flandmark%2Etxt%2Egz\",\n",
    "  \"GSE92742_Broad_LINCS_Level3_INF_mlr12k_n1319138x12328.gctx\"=\"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE92742&format=file&file=GSE92742%5FBroad%5FLINCS%5FLevel3%5FINF%5Fmlr12k%5Fn1319138x12328%2Egctx%2Egz\"\n",
    ")\n",
    "l1000_path = file.path(data_path, \"L1000\")\n",
    "dir.create(l1000_path, showWarnings = FALSE)\n",
    "for(filename in names(lincs_urls)){\n",
    "  download.file(lincs_urls[[filename]], file.path(l1000_path, filename))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d4a3fea",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Importing L1000 drug perturbations\n",
    "In this step we identify relevant drugs from the LINCS L1000 dataset.\n",
    "We read the L1000 perturbation information file and select those drugs that have known targets mapped on the co-expression disease network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b765a0bf",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "L1000_GSE92742_pert_info <- read.delim(file.path(l1000_path, \"GSE92742_Broad_LINCS_pert_info.txt\"), header=TRUE, quote = \"\")\n",
    "toRemove <- c(1,3,4,5,6,8)\n",
    "L1000_GSE92742_pert_info <- L1000_GSE92742_pert_info[, -toRemove]\n",
    "L1000_GSE92742_pert_info$pert_iname <- sapply(L1000_GSE92742_pert_info$pert_iname, tolower)\n",
    "\n",
    "map_drugs_on_graph$dat.drug.molecule_name <- gsub(map_drugs_on_graph$dat.drug.molecule_name, pattern = \".\", replacement = \"-\", fixed = TRUE)\n",
    "filter_by_L1000_drugs <- map_drugs_on_graph[which(map_drugs_on_graph$dat.drug.molecule_name %in% L1000_GSE92742_pert_info$pert_iname), ]\n",
    "filter_by_L1000_drugs$canonical_smiles <- NA\n",
    "filter_by_L1000_drugs$canonical_smiles <- L1000_GSE92742_pert_info$canonical_smiles[match(filter_by_L1000_drugs$dat.drug.molecule_name, L1000_GSE92742_pert_info$pert_iname)]\n",
    "filter_by_L1000_drugs <- unique(filter_by_L1000_drugs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12973b76",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Selecting Cell Type and Filtering L1000 Metadata\n",
    "We filter the signature metadata to include only entries where the experimental conditions match the ones of interest: `cell_id` is FIBRNPC, `pert_idose` is \"10 µM\", `pert_itime` is \"24 h\", and `pert_iname` is present in `filter_by_L1000_drugs$dat.drug.molecule_name`.\n",
    "We do a similar filtering for the control samples, with the exception that in this case the type of perturbation is set to \"DMSO\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77f8b2aa",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cell_type <- \"FIBRNPC\"\n",
    "\n",
    "col_meta_path_2 <- file.path(l1000_path, \"GSE92742_Broad_LINCS_sig_info.txt\")\n",
    "col_meta_2 <- read.delim(col_meta_path_2, sep = \"\\t\", stringsAsFactors = F)\n",
    "col_path_2 <- file.path(l1000_path, \"GSE92742_Broad_LINCS_Level3_INF_mlr12k_n1319138x12328.gctx\")\n",
    "\n",
    "col_meta_filtered_2 <- col_meta_2 %>%\n",
    "  dplyr::filter(cell_id==cell_type, pert_idose==\"10 µM\", pert_itime==\"24 h\") %>%\n",
    "  dplyr::filter(pert_iname %in% filter_by_L1000_drugs$dat.drug.molecule_name) \n",
    "\n",
    "col_meta_filtered_2_CTRL <- col_meta_2 %>%\n",
    "  dplyr::filter(cell_id==cell_type, pert_itime==\"24 h\", pert_iname==\"DMSO\")\n",
    "\n",
    "col_meta_filtered_2[, c(6,7,9,10)] <- NULL\n",
    "col_meta_filtered_2_CTRL[, c(6,7,9,10)] <- NULL\n",
    "\n",
    "col_meta_filtered_2 <- tidyr::separate_rows(col_meta_filtered_2, distil_id, sep = \"\\\\|\")\n",
    "col_meta_filtered_2 <- unique(col_meta_filtered_2)\n",
    "rownames(col_meta_filtered_2) <- 1:nrow(col_meta_filtered_2)\n",
    "\n",
    "col_meta_filtered_2_CTRL <- tidyr::separate_rows(col_meta_filtered_2_CTRL, distil_id, sep = \"\\\\|\")\n",
    "rownames(col_meta_filtered_2_CTRL) <- 1:nrow(col_meta_filtered_2_CTRL)\n",
    "col_meta_filtered_2_CTRL <- col_meta_filtered_2_CTRL[-which(col_meta_filtered_2_CTRL$pert_idose!=\"-666\"),]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69a34c1a",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Loading L1000 expression data\n",
    "We are ready to extract the expression data for our samples of interest from the Level 3 of the L1000 dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e850a71",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "GSE92742_ds <- parse.gctx(col_path_2, cid = col_meta_filtered_2$distil_id)\n",
    "GSE92742_CTRL <- parse.gctx(col_path_2, cid=unique(col_meta_filtered_2_CTRL$distil_id))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "144f37d6",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "We read the file \"GSE92742_Broad_LINCS_gene_info_delta_landmark.txt\" to obtain information about landmark genes and store the gene IDs in the `landmark_genes` variable as character strings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "442ef1c8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "landmark_genes <- read.delim(file.path(l1000_path, \"GSE92742_Broad_LINCS_gene_info_delta_landmark.txt\"), header=TRUE, quote=\"\")\n",
    "landmark_genes <- as.character(landmark_genes$pr_gene_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9baa12d4",
   "metadata": {},
   "source": [
    "Finally, we create an expression matrix for each drug using the `build_drug_expression_matrices` function. \n",
    "This function takes the L1000 data, L1000 control data, L1000 metadata, L1000 control metadata, a destination path, and the landmark genes as input.\n",
    "The resulting expression matrices, one for each drug, are saved in the \"./Expression_matrices_L1000\" directory in tab-delimited format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a305fa5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "create_L1000_exprmat <- build_drug_expression_matrices(L1000_data = GSE92742_ds, L1000_data_CTRL = GSE92742_CTRL, L1000_metadata = col_meta_filtered_2, L1000_metadata_CTRL = col_meta_filtered_2_CTRL, include_controls = TRUE, dest_path = drug_expr_path, inputgenes = landmark_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c07e446c",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Random control samples selection\n",
    "\n",
    "In this step, we will balance the number of samples between treatment and control groups when creating the expression matrices for each drug.\n",
    "The reason behind this approach is to avoid an imbalance in the number of columns, where the treatment matrices would have much fewer columns than the control matrices.\n",
    "Such an imbalance could lead to biased network analysis and produce highly similar networks for all drugs.\n",
    "\n",
    "To address this issue, we will randomly select control samples to match the number of samples in the treatment group for each drug.\n",
    "This balancing step is essential to prevent the dominance of control samples and facilitate meaningful comparisons between treatment and control conditions.\n",
    "\n",
    "We iterate over the files in the \"Expression_matrices_L1000/\" directory and for each expression matrix, we will select a random set of control samples equal in number to the treatment samples.\n",
    "The modified matrix is then saved back to the same file location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "586a1b57",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "myfiles <- list.files(drug_expr_path, pattern = \"Expression_matrix\")\n",
    "\n",
    "for (i in 1:length(myfiles)){\n",
    "  random_CTRL_matrix <- data.frame()\n",
    "  file <- read.table(file.path(drug_expr_path, myfiles[i]), header = TRUE, sep = \"\\t\", quote = \"\", row.names = 1)\n",
    "  CTRL <- grep(colnames(file), pattern = \"CTRL\", fixed = TRUE)\n",
    "  random_CTRL_matrix <- file[, -CTRL]\n",
    "  random_CTRL_matrix <- cbind(random_CTRL_matrix, file[, sample(x = CTRL, size = ncol(random_CTRL_matrix))])\n",
    "  write.table(random_CTRL_matrix, file = file.path(drug_expr_path, myfiles[i]), quote = FALSE, sep = \"\\t\", row.names=TRUE, col=NA)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "664b9c48",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Creating Drug Co-Expression Matrices\n",
    "\n",
    "In this step, we will create drug co-expression matrices using the perturbation profiles of the drugs in the L1000 dataset, similar to how we constructed the disease network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c4d7b08",
   "metadata": {
    "lines_to_next_cell": 2,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "myfiles <- list.files(drug_expr_path, pattern = \"Expression_matrix\")\n",
    "\n",
    "inputinform_drugs <- sapply(strsplit(myfiles, split = \"_\", fixed = TRUE), \"[\", 3)\n",
    "inputinform_drugs <- sapply(strsplit(inputinform_drugs, split = \".\", fixed = TRUE), \"[\", 1)\n",
    "\n",
    "for(i in 1:length(myfiles)){\n",
    "  file <- read.table(file.path(drug_expr_path, myfiles[i]), header = TRUE, sep = \"\\t\", quote = \"\", row.names = 1)\n",
    "  \n",
    "  generatematrices=get_ranked_consensus_matrix(gx_table = file, iMethods = c(\"clr\",\"aracne\",\"mrnet\"),\n",
    "                                               iEst = c(\"pearson\",\"spearman\",\"kendall\",\"mi.empirical\",\"mi.mm\",\"mi.shrink\",\"mi.sg\"),\n",
    "                                               iDisc=c(\"none\",\"equalfreq\",\"equalwidth\",\"globalequalwidth\"), ncores = 20, debug_output = TRUE, updateProgress = TRUE)\n",
    "  \n",
    "  \n",
    "  #Parse ranked matrix and get bin_mat and edge_rank\n",
    "  # Get edge rank list and binary inference matrix from edge rank matrix computed by get_ranked_consensus_matrix().\n",
    "  # parse_edge_rank_matrix parses the edge rank matrix created by using the internal function get_ranked_consensus_matrix_matrix() to get a ranked edge list and a binary matrix.\n",
    "  \n",
    "  rankMat.parsed=parse_edge_rank_matrix(edge_rank_matrix = generatematrices, edge_selection_strategy = \"default\",\n",
    "                                        mat_weights = \"rank\", topN = 10, debug_output = TRUE, updateProgress = TRUE)\n",
    "  \n",
    "  drug_conGraph <- get_iGraph(rankMat.parsed$bin_mat)\n",
    "  saveRDS(drug_conGraph, file=file.path(drug_expr_path, paste0(\"conGraph_\", inputinform_drugs[i], \".rds\")))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce4171c4",
   "metadata": {},
   "source": [
    "## Computing Pairwise Distances among Drug Co-Expression Networks\n",
    "\n",
    "Here, we utilize the `nettools` package to compute the distances between pairs of drug co-expression networks.\n",
    "The `netdist()` function takes the drug co-expression networks stored in the `infer_L1000_coexpr_networks` list as input.\n",
    "We specify the distance metric as \"HIM\", which stands for Hamming and Ipsen-Mikhailov, a measure for quantifying network differences.\n",
    "\n",
    "The computed distances are stored in the `ddist` object.\n",
    "We assign meaningful row and column names to the distance matrix, matching them with the drug names converted to lowercase.\n",
    "\n",
    "Finally, we construct a network that represents the distances between drugs based on their perturbational differences in gene expression profiles.\n",
    "This network offers a comprehensive view of the relationships and similarities among the drugs.\n",
    "The resulting `moa` object represents the drug distance network as a dataframe, where each row and column corresponds to a drug, and the cell values indicate the distance between them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c9ac205",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "myfiles <- list.files(path = drug_expr_path, pattern = \"conGraph\")\n",
    "\n",
    "# load drug co-expression netoworks based on the perturbational profiles of the 1000 landmark genes of L1000\n",
    "infer_L1000_coexpr_networks <- list()\n",
    "for (i in 1:length(myfiles)){\n",
    "  file <- readRDS(file.path(drug_expr_path, myfiles[i]))\n",
    "  infer_L1000_coexpr_networks[[i]] <- file\n",
    "}\n",
    "names(infer_L1000_coexpr_networks) <- sapply(strsplit(myfiles, \".\", fixed=TRUE), \"[\", 1)\n",
    "names(infer_L1000_coexpr_networks) <- sapply(strsplit(names(infer_L1000_coexpr_networks), \"_\", fixed=TRUE), \"[\", 2)\n",
    "\n",
    "# compute the pairwise distances among the drug co-expression networks\n",
    "ddist <- nettools::netdist(infer_L1000_coexpr_networks, d=\"HIM\", n.cores=4)\n",
    "mat_dist_mat <- ddist$HIM\n",
    "rownames(mat_dist_mat) <- colnames(mat_dist_mat) <- tolower(names(infer_L1000_coexpr_networks))\n",
    "# network of drug distances based on perturbational differences\n",
    "moa <- distance_to_dataframe(mat_dist_mat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b5b0543",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Computing Drug Distances based on Structural Similarity\n",
    "In this step, we retrieve the SMILES representations of the drugs from the PubChem service.\n",
    "We use the `fetch_smiles()` function to retrieve the SMILES representations of the drugs.\n",
    "The function takes the drug names (obtained from the row names of the distance matrix `mat_dist_mat`) as input and queries the PubChem service to fetch the corresponding SMILES strings.\n",
    "Next, we utilize the `chemicals_distance()` function to compute the drug distances.\n",
    "The function takes the drug SMILES representations stored in the drugs_smiles object as input.\n",
    "We specify the distance calculation method as \"fingerprint\", which utilizes molecular fingerprints to quantify the structural similarity between compounds.\n",
    "Other options include \"mcs\" distance which is based on the maximal common substructure and \"levenshtein\" which is based on the edit distance of the SMILES strings.\n",
    "The `python_bin` parameter represents the path to the Python interpreter, which is necessary for executing the underlying chemical similarity calculation code.\n",
    "The resulting `smiles` object represents the pairwise distances between drugs based on their structural similarity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01da5795",
   "metadata": {
    "lines_to_next_cell": 2,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# retrieve drug SMILES from pubchem service\n",
    "drugs_smiles <- fetch_smiles(rownames(mat_dist_mat))\n",
    "# drug distances based on structural similarity\n",
    "smiles <- chemicals_distance(drugs_smiles, method=\"fingerprint\", python=python_bin)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb295e5d",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "## Computing Drug Distances based on Target Connectivity\n",
    "In this part, we calculate the pairwise distances between drugs based on the connectivity of their respective targets in the disease network.\n",
    "The connectivity is determined by the average shortest path length between the targets in the network.\n",
    "\n",
    "We utilize the `get_avg_shortest_path()` function to compute the average shortest path length between the targets of each drug in the disease network represented by the `conGraph` object.\n",
    "The function takes the disease network graph (`Graph`), the drug-target associations (`drug_target_df`), and the column indices specifying the drug and target columns in the associations data frame (`drugs_col` and `targets_col`, respectively).\n",
    "\n",
    "The resulting `avg_drug_shortest_path` object contains the average shortest path lengths between the drugs based on their corresponding targets in the disease network.\n",
    "We then convert this information into a data frame using the `distance_to_dataframe()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b114074a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# search relevant drug targets\n",
    "opentargets_idx <- map_drugs_on_graph$dat.drug.molecule_name %in% tolower(names(drugs_smiles))\n",
    "filter_by_L1000_drugs <- unique(map_drugs_on_graph[opentargets_idx, ])\n",
    "targets <- filter_by_L1000_drugs\n",
    "# drug distances based on the shortest path length connecting their respective targets in the disease network\n",
    "avg_drug_shortest_path <- get_avg_shortest_path(Graph = conGraph, drug_target_df = filter_by_L1000_drugs, drugs_col = 6, targets_col = 1)\n",
    "graph <- distance_to_dataframe(avg_drug_shortest_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e532f171",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# Drug Combination Analysis with a Multi-Objective Genetic Algorithm\n",
    "\n",
    "In this step, we perform a drug combination analysis using a multi-objective genetic algorithm.\n",
    "The genetic algorithm takes into account multiple drug properties, including structural similarity, perturbational differences, network distances, and target connectivity.\n",
    "\n",
    "We initiate the genetic algorithm by calling the `genetic_algorithm()` function.\n",
    "This function takes several input parameters:\n",
    "\n",
    "    `smiles`: The drug structural distance calculated based on the drug SMILES retrieved earlier.\n",
    "    `moa`: The drug perturbational distance calculated based on the drug co-expression networks.\n",
    "    `graph`: The drug network distances based on the shortest path lengths connecting their respective targets in the disease network.\n",
    "    `disease_network`: The disease network used to incorporate information about the disease in the drug combination analysis.\n",
    "    `rank`: The ranked gene list based on the differential expression analysis of the disease.\n",
    "    `targets`: The relevant drug targets obtained from the OpenTargets database.\n",
    "\n",
    "The other parameters specified in the function control the behavior of the genetic algorithm:\n",
    "\n",
    "    `population_size`: The number of individuals in the initial population.\n",
    "    `n_offsprings`: The number of offspring produced in each generation through crossover and mutation.\n",
    "    `attribute_init_prob`: The probability of initializing each attribute (drug) in the population.\n",
    "    `attribute_mutation_prob`: The probability of mutating each attribute (drug) in the population.\n",
    "    `crossover_prob`: The probability of performing crossover between parent individuals.\n",
    "    `mutation_prob`: The probability of mutating individual attributes.\n",
    "    `n_generations`: The number of generations for which the genetic algorithm will run.\n",
    "    `python`: The path to the Python executable used to run the genetic algorithm.\n",
    "\n",
    "By running the genetic algorithm, we aim to find optimal drug combinations that optimize multiple objectives, such as structural differences, perturbational differences, network distances, and target connectivity.\n",
    "The algorithm explores the solution space, iteratively evolving the population of drug combinations over multiple generations to converge towards the best solutions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "011df5f4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#  returns a named list containing the following fields:\n",
    "#  drug_names: a list with the names of the drugs considered\n",
    "#  hall_of_fame: the list of best drug combinations identified during the execution of the algorithm, with their scores from the fitness function\n",
    "#  logbook: the average score of each fitness function for each generation\n",
    "#  population: a list of binary arrays representing the individuals in the population in the last generation of the algorithm.\n",
    "#\n",
    "#  The precomputed input data for the genetic algorithm can be loaded with\n",
    "# smiles = read.table(file.path(intermediate_path, \"smiles_dist.csv\"), sep=\",\", header=TRUE)\n",
    "# moa = read.table(file.path(intermediate_path, \"moa_dist.csv\"), sep=\",\", header=TRUE)\n",
    "# graph = read.table(file.path(intermediate_path, \"path_dist.csv\"), sep=\",\", header=TRUE)\n",
    "# rank = read.table(file.path(intermediate_path, \"graph_rank.csv\"), sep=\",\", header=TRUE)\n",
    "# targets = read.table(file.path(intermediate_path, \"drug_targets.csv\"), sep=\",\", header=TRUE)\n",
    "ga = genetic_algorithm(smiles, moa, graph, disease_network, rank, targets,\n",
    "                        population_size=100,\n",
    "                        n_offsprings=20,\n",
    "                        attribute_init_prob=0.3,\n",
    "                        attribute_mutation_prob=0.1,\n",
    "                        crossover_prob=0.7,\n",
    "                        mutation_prob=0.3,\n",
    "                        n_generations=500,\n",
    "                        python=python_bin\n",
    ")\n",
    "#  the logbook and the hall_of_fame of the run are stored as intermediate files run001_hof.json and run001_logbook.json"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "afbf4d09",
   "metadata": {},
   "source": [
    "The `genetic_algorithm` function returns a named list with the following fields:\n",
    "\n",
    "  * `drug_names`: A list containing the names of the drugs considered.\n",
    "  * `hall_of_fame`: The list of the best drug combinations identified throughout the algorithm's execution, accompanied by their corresponding scores from the fitness function.\n",
    "  * `logbook`: A record of the average score of each fitness function for every generation.\n",
    "  * `population`: A list of drug names representing the individuals in the population during the final generation of the algorithm.\n",
    "\n",
    "In the Jupyter notebook \"genetic_algorithm_results_exploration.ipynb\" we provide an exemplary demonstration of how to interpret the outcomes of the genetic algorithm in the context of our specific case study."
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
